lib-gphotos/client.go

```diff
@@ -57,31 +56,97 @@ func NewClient(oauthHTTPClient *http.Client, maybeToken ...*oauth2.Token) (*Clie
+		}
+		defer res.Body.Close()
+
+		// Get upload status
+		status := res.Header.Get("X-Goog-Upload-Status")
+		log.Printf("Status of upload URL '%s' is '%s'\n", *uploadURL, status)
+		if status == "active" {
+			offset, err = strconv.ParseInt(res.Header.Get("X-Goog-Upload-Size-Received"), 10, 0)
+			if err == nil && offset > 0 && offset < fileSize {
+				// Skip already uploaded part of the file
+				r.Seek(offset, io.SeekStart)
+			}
+		} else {
+			// Other known statuses "final" and "cancelled" are both considered an Error by the official Ruby client
+			// https://github.com/googleapis/google-api-ruby-client/blob/0.30.2/lib/google/apis/core/upload.rb#L250
+			// Let's restart the upload from scratch
+			*uploadURL = ""
+		}
+	}
+
+	if *uploadURL == "" {
+		// Get new upload URL
+		log.Printf("Getting new upload URL for '%s'\n", filename)
+		req, err := http.NewRequest("POST", fmt.Sprintf("%s/%s/uploads", basePath, apiVersion), nil)
+		if err != nil {
+			return "", err
+		}
+		req.Header.Set("Content-Length", "0")
+		req.Header.Set("X-Goog-Upload-Command", "start")
+		req.Header.Add("X-Goog-Upload-Content-Type", "application/octet-stream")
+		req.Header.Set("X-Goog-Upload-Protocol", "resumable")
+		req.Header.Set("X-Goog-Upload-Raw-Size", fmt.Sprintf("%d", fileSize))
+
+		res, err := c.Client.Do(req)
+		if err != nil {
+			return "", err
+		}
+		defer res.Body.Close()
+
+		// Read upload url
+		*uploadURL = res.Header.Get("X-Goog-Upload-URL")
+	}
+
+	log.Printf("Uploading content to '%s'\n", *uploadURL)
+
+	contentLength := int(fileSize - offset)
+	req, err := http.NewRequest("POST", *uploadURL, &ReadProgressReporter{r: r, max: contentLength, fileSize: int(fileSize)})
 	if err != nil {
+		log.Printf("Failed to prepare request: Error '%s'\n", err)
 		return "", err
 	}
-	req.Header.Set("Content-Type", "application/octet-stream")
-	req.Header.Add("X-Goog-Upload-File-Name", filename)
-	req.Header.Set("X-Goog-Upload-Protocol", "raw")
+	req.Header.Set("Content-Length", fmt.Sprintf("%d", contentLength))
+	req.Header.Add("X-Goog-Upload-Command", "upload, finalize")
+	req.Header.Set("X-Goog-Upload-Offset", fmt.Sprintf("%d", offset))
 
-	res, err := client.Client.Do(req)
+	res, err := c.Client.Do(req)
 	if err != nil {
+		log.Printf("\nFailed to process request '%s'\n", err)
 		return "", err
 	}
 	defer res.Body.Close()
 
 	b, err := ioutil.ReadAll(res.Body)
 	if err != nil {
+		log.Printf("Failed to read response '%s'\n", err)
 		return "", err
 	}
-	uploadToken := string(b)
+	uploadToken = string(b)
 	return uploadToken, nil
 }
 
-// Upload actually uploads the media and activates it on google photos
-func (client *Client) UploadFile(filePath string, pAlbumID ...string) (*photoslibrary.MediaItem, error) {
+// UploadFile actually uploads the media and activaes it on google photos
+// The uploadURL parameter will be updated
+func (c *Client) UploadFile(filePath string, uploadURL *string, pAlbumID ...string) (*photoslibrary.MediaItem, error) {
```

Comment:
Once again we are breaking the contract... This is gonna be a major change. I don't know if makes sense to maintain the old one for a while... What do you think?

Reply:
This is indeed a breaking change, I can add a new method indeed.

Reply:
That should be resolved now.

Reply:
Fixed for good now, see https://github.com/gphotosuploader/google-photos-api-client-go/pull/5#discussion_r308546487.